File: backend/api-gateway/src/routes/voice-recognition.ts
import { Router, Request, Response } from 'express';
import { query } from '../db';
import axios from 'axios';

const router = Router();

// Azure Speaker Recognition configuration
const AZURE_SPEAKER_KEY = process.env.AZURE_SPEAKER_RECOGNITION_KEY || '';
const AZURE_SPEAKER_REGION = process.env.AZURE_SPEAKER_RECOGNITION_REGION || 'eastus';
const AZURE_SPEAKER_BASE_URL = `https://${AZURE_SPEAKER_REGION}.api.cognitive.microsoft.com/speaker`;
const CONFIDENCE_THRESHOLD = parseFloat(process.env.VOICE_RECOGNITION_CONFIDENCE_THRESHOLD || '0.75');

// Check if voice recognition is enabled
const isVoiceRecognitionEnabled = () => {
  return !!AZURE_SPEAKER_KEY;
};

/**
 * GET /api/voice-recognition/status
 * Check if voice recognition is enabled
 */
router.get('/status', async (req: Request, res: Response) => {
  res.json({
    enabled: isVoiceRecognitionEnabled(),
    confidence_threshold: CONFIDENCE_THRESHOLD,
    region: AZURE_SPEAKER_REGION,
  });
});

/**
 * GET /api/voice-recognition/profiles
 * Get all voice profiles from database
 */
router.get('/profiles', async (req: Request, res: Response) => {
  try {
    const result = await query(`
      SELECT
        id,
        customer_id,
        customer_name,
        phone_number,
        email,
        policy_number,
        enrollment_status,
        enrollment_audio_duration_seconds,
        enrolled_at,
        last_recognized_at,
        recognition_count,
        created_at
      FROM customer_voice_profiles
      ORDER BY created_at DESC
    `);

    res.json({
      status: 'success',
      data: result.rows,
    });
  } catch (error: any) {
    console.error('Error fetching voice profiles:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

/**
 * GET /api/voice-recognition/profiles/:customerId
 * Get voice profile by customer ID
 */
router.get('/profiles/:customerId', async (req: Request, res: Response) => {
  try {
    const { customerId } = req.params;

    const result = await query(`
      SELECT *
      FROM customer_voice_profiles
      WHERE customer_id = $1
    `, [customerId]);

    if (result.rows.length === 0) {
      return res.status(404).json({
        status: 'error',
        message: 'Voice profile not found',
      });
    }

    res.json({
      status: 'success',
      data: result.rows[0],
    });
  } catch (error: any) {
    console.error('Error fetching voice profile:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

/**
 * POST /api/voice-recognition/enroll
 * Create voice profile and enroll customer
 * Body: { customerId, customerName, phoneNumber, email, policyNumber, audioBase64 }
 */
router.post('/enroll', async (req: Request, res: Response) => {
  try {
    const { customerId, customerName, phoneNumber, email, policyNumber, audioBase64 } = req.body;

    if (!customerId || !customerName || !audioBase64) {
      return res.status(400).json({
        status: 'error',
        message: 'Missing required fields: customerId, customerName, audioBase64',
      });
    }

    if (!isVoiceRecognitionEnabled()) {
      // Mock enrollment for demo
      const mockProfileId = `mock-profile-${customerId}-${Date.now()}`;

      await query(`
        INSERT INTO customer_voice_profiles (
          customer_id,
          customer_name,
          phone_number,
          email,
          policy_number,
          azure_profile_id,
          enrollment_status,
          enrollment_audio_duration_seconds,
          enrolled_at
        ) VALUES ($1, $2, $3, $4, $5, $6, 'enrolled', 32, NOW())
        ON CONFLICT (customer_id) DO UPDATE SET
          customer_name = EXCLUDED.customer_name,
          phone_number = EXCLUDED.phone_number,
          email = EXCLUDED.email,
          policy_number = EXCLUDED.policy_number,
          enrollment_status = 'enrolled',
          enrolled_at = NOW(),
          updated_at = NOW()
      `, [customerId, customerName, phoneNumber, email, policyNumber, mockProfileId]);

      return res.json({
        status: 'success',
        message: 'Voice profile enrolled (mock mode)',
        data: {
          profile_id: mockProfileId,
          customer_id: customerId,
          enrollment_status: 'enrolled',
        },
      });
    }

    // Real Azure enrollment
    console.log(`Creating Azure voice profile for customer: ${customerId}`);

    // Step 1: Create profile
    const createProfileResponse = await axios.post(
      `${AZURE_SPEAKER_BASE_URL}/identification/v2.0/text-independent/profiles`,
      { locale: 'en-US' },
      {
        headers: {
          'Ocp-Apim-Subscription-Key': AZURE_SPEAKER_KEY,
          'Content-Type': 'application/json',
        },
      }
    );

    const azureProfileId = createProfileResponse.data.profileId;
    console.log(`Azure profile created: ${azureProfileId}`);

    // Step 2: Enroll with audio
    const audioBuffer = Buffer.from(audioBase64, 'base64');

    const enrollResponse = await axios.post(
      `${AZURE_SPEAKER_BASE_URL}/identification/v2.0/text-independent/profiles/${azureProfileId}/enrollments`,
      audioBuffer,
      {
        headers: {
          'Ocp-Apim-Subscription-Key': AZURE_SPEAKER_KEY,
          'Content-Type': 'audio/wav',
        },
        timeout: 60000,
      }
    );

    const enrollmentStatus = enrollResponse.data.enrollmentStatus;
    console.log(`Enrollment status: ${enrollmentStatus}`);

    // Step 3: Store in database
    await query(`
      INSERT INTO customer_voice_profiles (
        customer_id,
        customer_name,
        phone_number,
        email,
        policy_number,
        azure_profile_id,
        enrollment_status,
        enrollment_audio_duration_seconds,
        enrolled_at
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
      ON CONFLICT (customer_id) DO UPDATE SET
        azure_profile_id = EXCLUDED.azure_profile_id,
        enrollment_status = EXCLUDED.enrollment_status,
        enrolled_at = EXCLUDED.enrolled_at,
        updated_at = NOW()
    `, [
      customerId,
      customerName,
      phoneNumber,
      email,
      policyNumber,
      azureProfileId,
      enrollmentStatus.toLowerCase(),
      32, // approximate duration
      enrollmentStatus === 'Enrolled' ? new Date() : null,
    ]);

    res.json({
      status: 'success',
      message: 'Voice profile enrolled',
      data: {
        profile_id: azureProfileId,
        customer_id: customerId,
        enrollment_status: enrollmentStatus,
      },
    });
  } catch (error: any) {
    console.error('Error enrolling voice profile:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

/**
 * POST /api/voice-recognition/identify
 * Identify speaker from audio sample
 * Body: { sessionId, audioBase64 }
 */
router.post('/identify', async (req: Request, res: Response) => {
  try {
    const { sessionId, audioBase64 } = req.body;

    if (!audioBase64) {
      return res.status(400).json({
        status: 'error',
        message: 'Missing required field: audioBase64',
      });
    }

    const startTime = Date.now();

    // Get all enrolled profiles
    const profilesResult = await query(`
      SELECT azure_profile_id, customer_id, customer_name
      FROM customer_voice_profiles
      WHERE enrollment_status = 'enrolled'
    `);

    if (profilesResult.rows.length === 0) {
      return res.json({
        status: 'success',
        data: {
          recognized: false,
          message: 'No enrolled profiles to match against',
        },
      });
    }

    const enrolledProfiles = profilesResult.rows;

    if (!isVoiceRecognitionEnabled()) {
      // Mock recognition for demo - randomly pick a profile
      const mockProfile = enrolledProfiles[0];
      const mockConfidence = 0.94;

      const recognitionTimeMs = Date.now() - startTime;

      // Log event
      await query(`
        INSERT INTO voice_recognition_events (
          session_id,
          customer_id,
          voice_profile_id,
          recognized,
          confidence_score,
          audio_duration_seconds,
          recognition_time_ms
        )
        SELECT $1::uuid, $2, id, true, $3, 10, $4
        FROM customer_voice_profiles
        WHERE customer_id = $2
      `, [sessionId || null, mockProfile.customer_id, mockConfidence, recognitionTimeMs]);

      // Update session
      if (sessionId) {
        await query(`
          UPDATE sessions
          SET voice_recognized = true,
              recognized_customer_id = $1,
              recognition_confidence = $2
          WHERE room_id = $3
        `, [mockProfile.customer_id, mockConfidence, sessionId]);
      }

      return res.json({
        status: 'success',
        data: {
          recognized: true,
          customer_id: mockProfile.customer_id,
          customer_name: mockProfile.customer_name,
          confidence: mockConfidence,
          recognition_time_ms: recognitionTimeMs,
        },
      });
    }

    // Real Azure recognition
    const audioBuffer = Buffer.from(audioBase64, 'base64');
    const profileIds = enrolledProfiles.map(p => p.azure_profile_id).join(',');

    console.log(`Identifying speaker against ${enrolledProfiles.length} profiles`);

    const identifyResponse = await axios.post(
      `${AZURE_SPEAKER_BASE_URL}/identification/v2.0/text-independent/profiles/identifySingleSpeaker?profileIds=${profileIds}`,
      audioBuffer,
      {
        headers: {
          'Ocp-Apim-Subscription-Key': AZURE_SPEAKER_KEY,
          'Content-Type': 'audio/wav',
        },
        timeout: 60000,
      }
    );

    const recognitionTimeMs = Date.now() - startTime;
    const identifiedProfileId = identifyResponse.data.identifiedProfile?.profileId;
    const confidence = identifyResponse.data.identifiedProfile?.score || 0;

    if (identifiedProfileId && confidence >= CONFIDENCE_THRESHOLD) {
      // Find customer info
      const matchedProfile = enrolledProfiles.find(p => p.azure_profile_id === identifiedProfileId);

      if (matchedProfile) {
        // Log successful recognition
        await query(`
          INSERT INTO voice_recognition_events (
            session_id,
            customer_id,
            voice_profile_id,
            recognized,
            confidence_score,
            audio_duration_seconds,
            recognition_time_ms
          )
          SELECT $1, $2, id, true, $3, 10, $4
          FROM customer_voice_profiles
          WHERE customer_id = $2
        `, [sessionId, matchedProfile.customer_id, confidence, recognitionTimeMs]);

        // Update session
        if (sessionId) {
          await query(`
            UPDATE sessions
            SET voice_recognized = true,
                recognized_customer_id = $1,
                recognition_confidence = $2
            WHERE room_id = $3
          `, [matchedProfile.customer_id, confidence, sessionId]);
        }

        // Update profile stats
        await query(`
          UPDATE customer_voice_profiles
          SET last_recognized_at = NOW(),
              recognition_count = recognition_count + 1
          WHERE customer_id = $1
        `, [matchedProfile.customer_id]);

        res.json({
          status: 'success',
          data: {
            recognized: true,
            customer_id: matchedProfile.customer_id,
            customer_name: matchedProfile.customer_name,
            confidence: confidence,
            recognition_time_ms: recognitionTimeMs,
          },
        });
      } else {
        throw new Error('Matched profile not found in database');
      }
    } else {
      // Recognition failed
      await query(`
        INSERT INTO voice_recognition_events (
          session_id,
          recognized,
          confidence_score,
          audio_duration_seconds,
          recognition_time_ms,
          fallback_to_manual_auth
        ) VALUES ($1, false, $2, 10, $3, true)
      `, [sessionId, confidence, recognitionTimeMs]);

      res.json({
        status: 'success',
        data: {
          recognized: false,
          confidence: confidence,
          recognition_time_ms: recognitionTimeMs,
          fallback_to_manual_auth: true,
        },
      });
    }
  } catch (error: any) {
    console.error('Error identifying speaker:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

/**
 * DELETE /api/voice-recognition/profiles/:customerId
 * Delete voice profile (GDPR compliance)
 */
router.delete('/profiles/:customerId', async (req: Request, res: Response) => {
  try {
    const { customerId } = req.params;

    // Get profile info
    const profileResult = await query(`
      SELECT azure_profile_id
      FROM customer_voice_profiles
      WHERE customer_id = $1
    `, [customerId]);

    if (profileResult.rows.length === 0) {
      return res.status(404).json({
        status: 'error',
        message: 'Voice profile not found',
      });
    }

    const azureProfileId = profileResult.rows[0].azure_profile_id;

    // Delete from Azure (if enabled)
    if (isVoiceRecognitionEnabled() && !azureProfileId.startsWith('mock-')) {
      await axios.delete(
        `${AZURE_SPEAKER_BASE_URL}/identification/v2.0/text-independent/profiles/${azureProfileId}`,
        {
          headers: {
            'Ocp-Apim-Subscription-Key': AZURE_SPEAKER_KEY,
          },
        }
      );
    }

    // Delete from database
    await query(`
      DELETE FROM customer_voice_profiles
      WHERE customer_id = $1
    `, [customerId]);

    res.json({
      status: 'success',
      message: 'Voice profile deleted',
    });
  } catch (error: any) {
    console.error('Error deleting voice profile:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

/**
 * GET /api/voice-recognition/analytics
 * Get voice recognition analytics
 */
router.get('/analytics', async (req: Request, res: Response) => {
  try {
    const result = await query(`
      SELECT * FROM voice_recognition_analytics
      ORDER BY date DESC
      LIMIT 30
    `);

    res.json({
      status: 'success',
      data: result.rows,
    });
  } catch (error: any) {
    console.error('Error fetching voice recognition analytics:', error);
    res.status(500).json({
      status: 'error',
      message: error.message,
    });
  }
});

export default router;

File: backend/voicebot-engine/voice_recognition.py
#!/usr/bin/env python3
"""
Voice Recognition Service
Uses Azure Cognitive Services Speaker Recognition for voice biometrics
"""

import os
import io
import asyncio
import logging
from typing import Optional, Dict, List
from dataclasses import dataclass
import httpx
from datetime import datetime

logger = logging.getLogger(__name__)


@dataclass
class VoiceProfile:
    """Voice profile data"""
    profile_id: str
    customer_id: str
    customer_name: str
    enrollment_status: str
    enrolled_at: Optional[datetime] = None


@dataclass
class RecognitionResult:
    """Voice recognition result"""
    recognized: bool
    customer_id: Optional[str] = None
    confidence: float = 0.0
    profile_id: Optional[str] = None
    recognition_time_ms: int = 0
    error: Optional[str] = None


class VoiceRecognitionService:
    """
    Azure Cognitive Services Speaker Recognition
    Handles voice enrollment and identification
    """

    def __init__(
        self,
        subscription_key: Optional[str] = None,
        region: Optional[str] = None,
        confidence_threshold: float = 0.75
    ):
        self.subscription_key = subscription_key or os.getenv("AZURE_SPEAKER_RECOGNITION_KEY")
        self.region = region or os.getenv("AZURE_SPEAKER_RECOGNITION_REGION", "eastus")
        self.confidence_threshold = confidence_threshold

        if not self.subscription_key:
            logger.warning("Azure Speaker Recognition key not configured - voice recognition disabled")
            self.enabled = False
        else:
            self.enabled = True

        self.base_url = f"https://{self.region}.api.cognitive.microsoft.com/speaker"
        self.headers = {
            "Ocp-Apim-Subscription-Key": self.subscription_key,
            "Content-Type": "application/json"
        }

    async def create_profile(self, customer_id: str, customer_name: str) -> Optional[str]:
        """
        Create a new voice profile for a customer
        Returns Azure profile ID
        """
        if not self.enabled:
            logger.info(f"Voice recognition disabled - mock profile for {customer_id}")
            return f"mock-profile-{customer_id}"

        try:
            async with httpx.AsyncClient() as client:
                # Create independent profile (for speaker identification)
                response = await client.post(
                    f"{self.base_url}/identification/v2.0/text-independent/profiles",
                    headers=self.headers,
                    json={"locale": "en-US"}
                )

                if response.status_code == 201:
                    data = response.json()
                    profile_id = data.get("profileId")
                    logger.info(f"Created voice profile {profile_id} for customer {customer_id}")
                    return profile_id
                else:
                    logger.error(f"Failed to create profile: {response.status_code} - {response.text}")
                    return None

        except Exception as e:
            logger.error(f"Error creating voice profile: {e}")
            return None

    async def enroll_profile(
        self,
        profile_id: str,
        audio_data: bytes,
        customer_id: str
    ) -> bool:
        """
        Enroll voice profile with audio sample
        Audio should be: WAV format, 16kHz, mono, 16-bit PCM
        Minimum 4 seconds, recommended 30+ seconds
        """
        if not self.enabled:
            logger.info(f"Voice recognition disabled - mock enrollment for {customer_id}")
            return True

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                # Azure expects raw audio in request body
                headers = {
                    **self.headers,
                    "Content-Type": "audio/wav"
                }

                response = await client.post(
                    f"{self.base_url}/identification/v2.0/text-independent/profiles/{profile_id}/enrollments",
                    headers=headers,
                    content=audio_data
                )

                if response.status_code == 200:
                    data = response.json()
                    enrollment_status = data.get("enrollmentStatus")
                    logger.info(f"Enrollment for {profile_id}: {enrollment_status}")

                    # enrollmentStatus: "Enrolling" or "Enrolled"
                    return enrollment_status == "Enrolled"
                else:
                    logger.error(f"Enrollment failed: {response.status_code} - {response.text}")
                    return False

        except Exception as e:
            logger.error(f"Error enrolling voice profile: {e}")
            return False

    async def identify_speaker(
        self,
        audio_data: bytes,
        candidate_profile_ids: List[str]
    ) -> RecognitionResult:
        """
        Identify speaker from audio sample
        Returns recognition result with matched profile and confidence
        """
        start_time = datetime.now()

        if not self.enabled:
            logger.info("Voice recognition disabled - returning mock result")
            # For demo purposes, randomly recognize as first candidate
            if candidate_profile_ids:
                return RecognitionResult(
                    recognized=True,
                    profile_id=candidate_profile_ids[0],
                    confidence=0.94,
                    recognition_time_ms=150
                )
            return RecognitionResult(recognized=False)

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                # Azure expects profile IDs as query params
                profile_ids_param = ",".join(candidate_profile_ids)

                headers = {
                    **self.headers,
                    "Content-Type": "audio/wav"
                }

                response = await client.post(
                    f"{self.base_url}/identification/v2.0/text-independent/profiles/identifySingleSpeaker?"
                    f"profileIds={profile_ids_param}",
                    headers=headers,
                    content=audio_data
                )

                end_time = datetime.now()
                recognition_time_ms = int((end_time - start_time).total_seconds() * 1000)

                if response.status_code == 200:
                    data = response.json()
                    identified_profile_id = data.get("identifiedProfile", {}).get("profileId")
                    confidence = data.get("identifiedProfile", {}).get("score", 0.0)

                    if identified_profile_id and confidence >= self.confidence_threshold:
                        logger.info(f"Speaker identified: {identified_profile_id} (confidence: {confidence})")
                        return RecognitionResult(
                            recognized=True,
                            profile_id=identified_profile_id,
                            confidence=confidence,
                            recognition_time_ms=recognition_time_ms
                        )
                    else:
                        logger.info(f"Speaker not recognized (confidence too low: {confidence})")
                        return RecognitionResult(
                            recognized=False,
                            confidence=confidence,
                            recognition_time_ms=recognition_time_ms
                        )
                else:
                    error_msg = f"Identification failed: {response.status_code}"
                    logger.error(error_msg)
                    return RecognitionResult(
                        recognized=False,
                        error=error_msg,
                        recognition_time_ms=recognition_time_ms
                    )

        except Exception as e:
            error_msg = f"Error identifying speaker: {e}"
            logger.error(error_msg)
            return RecognitionResult(
                recognized=False,
                error=error_msg
            )

    async def delete_profile(self, profile_id: str) -> bool:
        """Delete a voice profile (GDPR compliance)"""
        if not self.enabled:
            return True

        try:
            async with httpx.AsyncClient() as client:
                response = await client.delete(
                    f"{self.base_url}/identification/v2.0/text-independent/profiles/{profile_id}",
                    headers=self.headers
                )

                if response.status_code == 204:
                    logger.info(f"Deleted voice profile: {profile_id}")
                    return True
                else:
                    logger.error(f"Failed to delete profile: {response.status_code}")
                    return False

        except Exception as e:
            logger.error(f"Error deleting voice profile: {e}")
            return False

    async def get_profile_status(self, profile_id: str) -> Optional[Dict]:
        """Get current status of a voice profile"""
        if not self.enabled:
            return {"enrollmentStatus": "Enrolled"}

        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/identification/v2.0/text-independent/profiles/{profile_id}",
                    headers=self.headers
                )

                if response.status_code == 200:
                    return response.json()
                else:
                    logger.error(f"Failed to get profile status: {response.status_code}")
                    return None

        except Exception as e:
            logger.error(f"Error getting profile status: {e}")
            return None


# Global service instance
_voice_recognition_service: Optional[VoiceRecognitionService] = None


def get_voice_recognition_service(
    subscription_key: Optional[str] = None,
    region: Optional[str] = None,
    confidence_threshold: float = 0.75
) -> VoiceRecognitionService:
    """Get or create voice recognition service singleton"""
    global _voice_recognition_service

    if _voice_recognition_service is None:
        _voice_recognition_service = VoiceRecognitionService(
            subscription_key=subscription_key,
            region=region,
            confidence_threshold=confidence_threshold
        )

    return _voice_recognition_service
